### ISSUE TITLE
FirebaseInstanceID Refetching default token infinite loop.

### ISSUE THREAD

**Issue by lauretmichiel1** [firebase/firebase-ios-sdk#259]
### Describe your environment

  * Xcode version: 8.3.3
  * Firebase SDK version: Firebase/Core (4.0.4)
  * Library version: FirebaseInstanceID (2.0.1) and FirebaseInstanceID (2.0.2)
  * Firebase Product: Auth, Database, Core, Messaging, RemoteConfig
### Describe the problem

I am running into an issue in two apps we are making. When starting the app for the first time the CPU is being hogged constantly at around ~40%-60%. When looking in the Time Profiler I noticed that it was something related to FIRInstanceID. When adding `-FIRDebugEnabled` in arguments passed on launch. The debug console gets spammed with these messages:  

```
2017-09-08 11:22:09.188 [92131] <Debug> [Firebase/InstanceID][I-IID003016] Received APNS token while fetching default token. Refetching default token.
2017-09-08 11:22:09.190 [92131] <Debug> [Firebase/InstanceID][I-IID003016] Received APNS token while fetching default token. Refetching default token.
2017-09-08 11:22:09.216 [92131] <Debug> [Firebase/InstanceID][I-IID003016] Received APNS token while fetching default token. Refetching default token.
2017-09-08 11:22:09.216 [92131] <Debug> [Firebase/InstanceID][I-IID003016] Received APNS token while fetching default token. Refetching default token.
2017-09-08 11:22:09.217 [92131] <Debug> [Firebase/InstanceID][I-IID003016] Received APNS token while fetching default token. Refetching default token.
2017-09-08 11:22:09.217 [92131] <Debug> [Firebase/InstanceID][I-IID003016] Received APNS token while fetching default token. Refetching default token.
```
![instanceidloop](https://user-images.githubusercontent.com/31561244/30372708-f82d184e-987e-11e7-8404-2509cc85dc7f.png)
![timeprofiler](https://user-images.githubusercontent.com/31561244/30372717-fbe44b7e-987e-11e7-895d-268c01252e45.png)

This doesn't stop until I kill and restart the app. However when we clean install the app the same problem occurs.

I have now managed to workaround this issue by adding pod 'FirebaseInstanceID', '2.0.0' to the Podfile. If I use pod 'FirebaseInstanceID', '2.0.1' or pod 'FirebaseInstanceID', '2.0.2' the problem occurs. Not including FirebaseInstanceID into the Podfile automatically installs version FirebaseInstanceID 2.0.2. Note that when commenting out all code related to registering for push notifications the above still occurs.

If you need more info please let me know.

Michiel

**Comment by katowulf** [comment]
Michiel, before we'll be able to make any progress, we'll need a simple and complete code sample to reproduce the problem and detailed steps to reproduce. That is required in the [issue template](https://github.com/firebase/firebase-ios-sdk/blob/master/.github/ISSUE_TEMPLATE.md). It may also help to provide the iOS version.

**Comment by rsattar** [comment]
Hey @lauretmichiel1 I've repro'ed the issue, and it's already been fixed for the next version of Firebase Instance ID. Very sorry about that! 

This is happening if: `registerForRemoteNotifications` is not called on first start and with the FCM SDK integrated. FYI `registerForRemoteNotifications` can be called without trigger user notification permissions. 

You can stay at the lower Pod version for now, but hopefully the fix should be available soon.

**Comment by rsattar** [comment]
For now, a quick workaround is to simply add:
```
application.registerForRemoteNotifications()
```
to your App Delegate.

**NOTE**: This is safe to call on app startup, as it does _not_ show the user notifications permissions dialog to the user. The dialog is only shown when you call `UNUserNotificationCenter.current().requestAuthorization(...)` or `application.registerUserNotificationSettings(...)`.

**Comment by lauretmichiel1** [comment]
@rsattar thank you for responding so quickly. Other question while we're at it, can not calling registerForRemoteNotifications have any influence on not registering for a valid APNS token in some cases? I do this at a later time when the user presses a button. With the same flow some iOS clients don't receive push messages from topics while others do, haven't been able to figure out why yet.

**Comment by rsattar** [comment]
> can not calling registerForRemoteNotifications have any influence on not registering for a valid APNS token in some cases?

Yeah, calling `registerForRemoteNotifications()` is what triggers iOS to give you an APNs token, so if you're not calling it, then iOS doesn't give an APNs token, which also means that the FCM SDK won't get an APNs token. FCM creates an FCM Token regardless of APNs tokens, currently, and will "enhance" the FCM token with an APNs token, once the APNs token is provided by iOS.

It's totally normal (and actually better!) to ask for the user's permissions to send them notifications later in the onboarding process, because the chances of them accepting are [much, much higher](https://library.launchkit.io/the-right-way-to-ask-users-for-ios-permissions-96fa4eb54f2c).

However, you can still call `registerForRemoteNotifications()` as soon as your app starts, and wait to call `registerUserNotificationSettings(...)` or `UNUserNotificationCenter`'s `requestAuthorization` method later on, when it's appropriate. That will ensure that you always get an APNs token (which FCM can work with) early, and your app can now received silent notifications immediately (e.g. `content-available: true`). When your users grant you permissions later, iOS will again give you the APNs token, but you'll have already associated it.

Hope that helps!

**Comment by paulb777** [comment]
Fixed in Firebase 4.2.0 FirebaseInstanceID 2.0.3 release today.
